Added CrossSubjectContext and refactored MotorImageryContexts

bibliolytic opened the pull request:
This isn't the finished version but I wanted to begin the pull request now since I think there are some more changes in order.

I added the CrossSubject context, as well as the MotorImagery contexts into a base and inheriting classes. However, I don't like the current more flat hierarchy where each usable class (e.g. MotorImageryTwoClassWithinSubject) needs to inherit from multiple classes and then virtually re-use the same code. I still think we should have the paradigm-related classes (e.g. MotorImageryTwoClasses/MotorImageryMultipleClasses) take as an argument an instance of BaseContext. 

I further noticed that the interfaces weren't being enforced so I added some abstractmethod decorators to force inheriting classes to implement the minimum necessary methods. 

Lastly, this does not have the SubjectUpdateContext because that is not sklearn compatible, I believe. Sklearn allows for data-driven feature generation/alteration, so domain adaptation, but methods that work in the feature space would essentially take some data and return an instance of BaseClassifier, instead of the (X,y) that the pipeline assumes. Unless I'm missing something, there is no pipeline-compatible way of doing this?

alexandrebarachant commented on moabb/contexts/base.py:
```diff
@@ -44,7 +46,29 @@ def __init__(self, datasets, pipelines):
+        Output:
+            X:       ndarray (trials, channels, timepoints) of data
+            y:       ndarray (trials,) **1 or 2d** of labels
+            groups:  ndarray (trials,) specifying which subject each trial belongs to
```
groups can represent any kind of information, i.e. runs, sessions, subjects.

bibliolytic replied:
Given that there is a list of subjects this is an issue, right? That needs to be preserved in X an y as a general rule, I think? So we could alter the signature in all classes such that X and y and groups are lists where each member corresponds to the data of the corresponding member of subjectlist?

alexandrebarachant replied:
No, this is not a bug.
For example, with your PR you reverted the fact that for many dataset i used session for group index in order to do a leave-one session out CV.

we can not enforce group to be subjects, it could be anything that is relevant for the CV.

bibliolytic replied:
Gotcha. In that case I'll change the signature of prepare_data to return lists of X, y, and groups, such that both subject specificity and the group variable are preserved

alexandrebarachant commented on moabb/contexts/motor_imagery.py:
```diff
@@ -1,47 +1,35 @@
+
+        """
         raws = dataset.get_data(subjects=subjects)
         raws = raws[0]
```
this was a leftover from single subject context.
`get_data()` return an array of array of raw. 
the first level correspond to the subject,
the second to each 'continuous' recording (i.e. could be a run or a session)

alexandrebarachant commented:
Okay, i will try to review that tonight.
Thanks a lot :+1:

alexandrebarachant commented on moabb/contexts/motor_imagery.py:
```diff
@@ -1,47 +1,35 @@
-        """epoch data"""
+        """epoch data
+
+        NOTE: Is this necessary to have defined within the 'motor imagery'
```
Indeed. Initially, my intention was to fix parameter in the code (i.e. not having them in arguments) to avoid performance hacking by changing parameters. But now this can probably be reused across many paradigm.

this preprocessing is kind of standard, but i'm wondering if it's not too limiting. But we still have the option to redifine it in the context.

alexandrebarachant commented on moabb/contexts/motor_imagery.py:
```diff
@@ -76,32 +59,62 @@ def prepare_data(self, dataset, subjects):
+            if equalize:
+                equalize_epoch_counts(epochs_list)
             ep = concatenate_epochs(epochs_list)
             groups.extend([ii] * len(ep))
```
With this, groups corresponds to session / runs, not subjects

bibliolytic commented:
I've updated things and implemented one solution to the SubjectUpdate context we were discussing, in which something is built on top of the sklearn pipeline that allows for a 'pre_fit' method or 'pre_transform' method that sets parameters for the fit method. I also changed the 'info' to be a DataFrame that specifies, for each data matrix, both which subject and which recording session.

alexandrebarachant commented:
are you happy with this PR ? what's the status ?

bibliolytic commented:
I'm taking it off--it's been too long, I don't remember the code, and I don't see it doing what I want. Will make another one when I get back to this level.
